users/vieting/experiments/switchboard/ctc/feat/baseline_args.py

```diff
@@ -237,6 +240,47 @@ def get_returnn_config(
         # network["source"] = specaug_layer_jingjing(in_layer=["features"])
         pass
 
+    if recognition:
+        preemphasis_perturbation = False
+        codec_perturbation = False
+        # If audio perturbation is enabled, we need to add the perturbation layers to the feature network
+        audio_perturb_args = extra_args.get("audio_perturb_args", {})
+        preemphasis_perturbation = "preemphasis" in audio_perturb_args
+        codec_perturbation = "codecs" in audio_perturb_args
+        extra_args.pop("audio_perturb_args", None)
+        extra_args.pop("audio_perturb_runner", None)
+
+    if audio_perturbation and recognition:
+        if preemphasis_perturbation:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == "data":
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+            audio_perturb_args = extra_args.get("audio_perturb_args", {})
+            default_coeff = audio_perturb_args.get("default_coeff", 0.97)
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=default_coeff).get_as_subnetwork()
+        elif codec_perturbation:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == "data":
+                    feature_net["subnetwork"][layer]["from"] = "codec"
+            feature_net["subnetwork"]["codec"] = {
+                "class": "subnetwork",
+                "from": ["data"],
+                "subnetwork": {
+                    "magnitude": {
+                        "class": "eval",
+                        "from": "data",
+                        "eval": "tf.math.log1p(255.0 * tf.abs(source(0))) / tf.math.log1p(255.0)",
+                    },
+                    "output": {
+                        "class": "eval",
+                        "from": "magnitude",
+                        "eval": "tf.sign(source(0)) * source(1)",
+                        "from": ["data", "magnitude"],
+                    },
+                },
+                "trainable": False,
+            }
```

Comment:
You can just put that in a single `EvalLayer` instead of a subnetwork. I think that's easier to read.

Reply:
With the addition of the assert_range layer, I think it can stay a subnetwork.